### PR TITLE
feat: accept a `File` interface, a subset of os.File

### PR DESCRIPTION
**Pull request by rvagg**
Allows a novel type to be passed that may do interesting things with output, such as teeing the write stream to a different location.

This is a new take on what #360 was trying to achieve. The problem with #360 is that it doesn't support full `Get()` operations, which we happen to need for some cases of DAGs when using graphsync. Instead, this accepts the fact that we need the full blockstore, so we'll write that out to a file, but since we only accept a subset of `os.File` we can replace it with another type that we can intercept and use to _stream_ the data elsewhere. For Lassie, we want to both have the blockstore, and be able to stream the contents of the CARv1 over the http interface, then clean up the blockstore when we're all done.

This only supports CARv1 because v2 requires a rewinding seek back to the header.

Is this too much of a hack?

### CODE REVIEW

File: v2/blockstore/readwrite.go [hunk 1]
```diff
@@ -33,14 +33,24 @@ var _ blockstore.Blockstore = (*ReadWrite)(nil)
...
+	f          File
 	dataWriter *internalio.OffsetWriteSeeker
 	idx        *insertionIndex
 	header     carv2.Header
 
 	opts carv2.Options
 }
 
+// File is a subset of os.File that is used by a ReadWrite
+type File interface {
```
[the remark this answers is not in the document]
Comment (rvagg) [reply v2/blockstore/readwrite.go:45]: Lassie needs a blockstore to read and write but it also needs to stream the CARv1 via HTTP to a caller. Unfortunately graphsync needs the ability to interrogate previously fetched blocks so we need something like the blockstore interface hanging around. So with this strategy we can give it a file but with the interface we can also tee the writer to an http stream.
It's not exactly pretty but I didn't want to have to reimplement ReadWrite all again. 

I'm not proud of this, it's very rough, but points toward what I'm trying to achieve: https://github.com/filecoin-project/lassie/pull/60

[the remark this answers is not in the document]
Comment (masih) [reply v2/blockstore/readwrite.go:45]: Sounds like the only concrete feature here is the need to tee the generated bytes onto other writers.

I have not looked too closely into alternative ways to achieve this; however, I wonder if a simpler interface would be sufficient _if_ we:
*  introduce a new constructor, called `OpenReadWriteStream` that takes the simpler interface, and
* remove support for resumption in the streaming variation.
  * `Stat`, `Truncate` and `Read` are all used by the resumtion functionality implemented in the constructor taking files.

Unless I have missed something, i think this interface would then suffice:

```go
type ReaderWriterCloserAt interface {
	io.ReaderAt
	io.WriterAt
	io.Closer
}
```

[the remark this answers is not in the document]
Comment (rvagg) [reply v2/blockstore/readwrite.go:45]: I considered that but the surgery of internals is **much** more invasive. There's even an io.Reader use in here. In this case the ReadWrite ends up taking responsibility for not being able to support certain features. At least with then interface approach it's caveat emptor and I take responsibility for a hobbled File implementation at my peril.

Before we go about wrestling ReadWrite into even more incoherence, I might try a different approach: https://github.com/filecoin-project/lassie/pull/60#issuecomment-1411840227

It'll mean a new simple (imperative) writable interface but that's been missing from v2 anyway so may not be a bad addition.

[the remark this answers is not in the document]
Comment (masih) [reply v2/blockstore/readwrite.go:45]: Sounds good; re `io.Reader` I think it's only used to read version during resumption.

**Comment by rvagg** [comment]
satisfied the needs behind this in the v2/storage package in v2.7.0

**Pull request closed without merging**